### PR TITLE
[feature](nereids): SimplifyCastRule

### PR DESCRIPTION
**Pull request by jackwener**
# Proposed changes

Issue Number: close #xxx

## Problem summary

- Remove redundant cast like
  - cast(1 as int) -> 1.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [x] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [x] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [x] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/SimplifyCastRule.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        }
+
+        // We forbidden subclass of NumericType convert to other subclass of NumericType,
+        if (!(left instanceof NumericType) || (!(right instanceof NumericType))) {
+            return true;
+        }
+
+        // It after remove redundant cast. so left.class != right.class
+        // So we don't need to return left.class == right.class
+        return false;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/SimplifyCastRule.java:95]:
```suggestion
        // We forbidden subclass of NumericType convert to other subclass of NumericType,
        if (left instanceof NumericType && right instanceof NumericType) {
            return false;
        }

        // It after remove redundant cast. so left.class != right.class
        // So we don't need to return left.class == right.class
        return true;
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/SimplifyCastRule.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    // It after remove redundant cast. so left.class != right.class
+    // We forbidden subclass of NumericType convert other subclass of NumericType,
+    // except subclass of IntegralType can mutual conversion.
+    private boolean check(DataType left, DataType right) {
+        // int, bigint ... can mutual conversion.
+        if (left instanceof IntegralType && right instanceof IntegralType) {
+            return true;
+        }
+
+        // We forbidden subclass of NumericType convert to other subclass of NumericType,
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rewrite/rules/SimplifyCastRule.java:88]: add some example to explain why forbidden this situation

**Comment by englefly** [comment]
The following two casts are not equivalent.

mysql> select cast(cast("1990-09-01 1:00:00" as datetime) as int);
+-----------------------------------------------------+
| CAST(CAST('1990-09-01 1:00:00' AS DATETIME) AS INT) |
+-----------------------------------------------------+
|                                         -1977439664 |
+-----------------------------------------------------+
1 row in set (0.01 sec)

mysql> select cast("1990-09-01 1:00:00" as int);
+-----------------------------------+
| CAST('1990-09-01 1:00:00' AS INT) |
+-----------------------------------+
|                              NULL |
+-----------------------------------+
1 row in set (0.01 sec)

**Review by 924060929 — APPROVED** [review]
LGTM

**Pull request merged**